File: SEPTEMBER 23/03 September/62. Unique Paths LC.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> uniquePath(m+1, vector<int>(n+1,0));
        uniquePath[m][n] = 1;
        for(int i=m; i>0; i--)
        {
            for(int j=n; j>0; j--)
            {
                if(j-1>0)
                    uniquePath[i][j-1]+=uniquePath[i][j];
                if(i-1>0)   
                    uniquePath[i-1][j]+=uniquePath[i][j];
            }
        }
        return uniquePath[1][1];
    }
};

File: APRIL 2025/05 April/AMZN/62. Unique Paths LC.java
class Solution {
    public int uniquePaths(int m, int n) {
        // I need to make (m-1) D moves and (n-1) R moves
        // In total I have (m-1+n-1)C(m-1) or (m-1+n-1)C(n-1) possible paths to take.

        // ans = [1*2*..*(m-1)m*(m+1)...(m+n-2)] / ([1*2*..(m-1)]*[1*2*..(n-1)]) 
        // ans = [m*(m+1)...(m+n-2)] / [1*2*..(n-1)] // assuming m>=n
        // when continuing this multiplication and division operation, the numerator will always be divisible by denominator;
        if(m==1 || n==1)
            return 1;
        m--;
        n--;
        if(m<n)
        {
            int tmp = m;
            m = n;
            n = tmp;
        }
        long ans = 1;

        for(int numerator=m+1, denominator=1; numerator<=(m+n); numerator++, denominator++) {
            ans *= numerator;
            ans /= denominator;
        }
        return (int)ans;
    }
}
